src/basic.jl

```diff
@@ -43,20 +48,25 @@ This should usually be an empty `Vector{S}` where `S` is the type of the state f
 """
 function particle_memory end
 
-function update(up::BasicParticleFilter, b::ParticleCollection, a, o)
+function update(up::BasicParticleFilter, b::AbstractParticleBelief, a, o)
     pm = up._particle_memory
     wm = up._weight_memory
     resize!(pm, n_particles(b))
     resize!(wm, n_particles(b))
+    if (calculate_ess(wm) < up.resampling_threshold)
+        resampled_particle_collection = resample(
+            up.resampler,
+            WeightedParticleBelief(pm, wm, sum(wm), nothing),
+            up.predict_model,
+            up.reweight_model,
+            b, a, o,
+            up.rng)
+		num_particles = n_particles(resampled_particle_collection)
+        return WeightedParticleBelief(resampled_particle_collection.particles, fill(1.0 / num_particles, num_particles))
+    end
```

Comment:
actually, if you resample before `reweight!`, the update function does not work as expected

Reply:
now resampling at the end

Reply:
I don't quite understand this decision. It seems much better to resample before prediction and I don't understand why it didn't work.

Reply:
> I don't quite understand this decision. It seems much better to resample before prediction and I don't understand why it didn't work.

I handled the edge case (first iteration) by checking `wm` is empty. The `resample` function should not be called when the `update` function is called for the first time